### DISCUSSION TITLE
Project 2: CUDA non-rigid registration

### DISCUSSION THREAD

**Discussion by deka27** [dipy/dipy#3129] [#GSoC/GSoD]
Hi @sreekarchigurupati @pjsjongsung @skoudoro ,

Hope everyone is doing well.

# Intro

I'm Kaustav Deka, a full stack developer based in India. While I'm deeply fascinated by graphics, 3D modeling, and related fields, I must humbly admit that my experience in CUDA programming is somewhat fundamental. Nonetheless, I've dabbled in CUDA and have a genuine interest in exploring its possibilities further. My journey in software development has exposed me to various technologies, and I'm always eager to learn and grow in areas that intrigue me.

# Work Proposal

I am extremely interested in applying for the CUDA non-rigid registration project with DIPY this summer. 

I am enthusiastic about elevating my skills to new heights by embracing the complexities of non-rigid registration.
This endeavor presents an opportunity for me to delve deeper into GPU data structures and optimization techniques.
Motivated by the prospect of driving meaningful advancements in medical imaging, I aspire to leverage my growing proficiency in CUDA programming. 
I am confident that my dedication and evolving expertise make me well-suited to contribute meaningfully to this open-source project as part of the Google Summer of Code program.

# Outro

Thank you for considering my application, and I look forward to the possibility of working together to make a meaningful impact in this field. 

Warm Regards,
Kaustav

**Comment by pjsjongsung** [comment]
Thank you for the interest @deka27 . We are excited that someone is interested in this topic. Hopefully @sreekarchigurupati can provide more help on specific project terms. Note that as a Python Software Foundation sub-org, we do require some kind of contribution to DIPY beforehand. So if you could look over some of the issues that we have right now, that would be awesome. Looking forward to your final proposal!

**Reply by deka27** [reply]
Hi @pjsjongsung, Thank you for the response. 
I have already submitted 1 PR towards DIPY #3128 . I am waiting for response on that PR. Meanwhile I am working on another PR, hopefully I can submit soon.

**Comment by deka27** [comment]
Hi, may I know what would be the next steps for the application. Do I submit my final proposal here?

Thank you.

**Reply by pjsjongsung** [reply]
No the final proposal should be submitted through the GSoC website, and we view it through there. You should be able to find our projects under Python Software Foundation.

**Comment by deka27** [comment]
Hi @pjsjongsung @skoudoro @sreekarchigurupati ,

This is my draft for the final proposal/application, please let me know if I am going in the right direction.

I got this template from the python-gsoc documentations. [Link](https://github.com/python-gsoc/python-gsoc.github.io/blob/main/ApplicationTemplate.md)

----------------

# DIPY: CUDA Acceleration for Non-Rigid Registration

## About me
1. Name: **Kaustav Deka**
2. Contact info - mailerdeka@gmail.com
3. Time zone - IST `UTC +5:30`
4. Github Link - [deka27](https://github.com/deka27)

## Code contribution
* Link to the pull request(s)
	* [PR #3132](https://github.com/dipy/dipy/pull/3132)

## Project information
1. Sub-org name: **DIPY**
   
2. Project Abstract:
This project aims to enhance the efficiency of neuroimaging preprocessing by implementing CUDA acceleration for non-rigid registration in DIPY, potentially extending to other algorithms such as tracking and segmentation, if time permits.

3. Detailed Description:
Non-rigid registration is crucial for accurate neuroimaging analysis but is often time-consuming due to its complexity. By integrating CUDA acceleration into DIPY's non-rigid registration process, this project seeks to significantly reduce processing times, facilitating faster and more efficient neuroimaging workflows.

4. Weekly timeline
   * **Community Bonding**: Review existing DIPY non-rigid registration algorithms, set up a CUDA development environment, and plan project milestones with mentors.
   * **Week 1 - 2**: Begin translating the current non-rigid registration algorithm from DIPY into a CUDA-compatible version, focusing on understanding the algorithm's complexity and identifying key areas for acceleration. Furthermore I could implement optimized memory management and parallel processing strategies to improve algorithm efficiency.
   * **Week 3**: Complete the initial CUDA implementation of the non-rigid registration algorithm. Begin benchmarking against the original algorithm to assess performance improvements.
   * **Week 4**: Analyze benchmark results and identify areas for further optimization. Apply advanced CUDA features to enhance algorithm performance.
   * **Week 5**: Expand testing to include various neuroimaging datasets, ensuring the algorithm's accuracy and performance consistency across different scenarios.
   * **Week 6**: Review progress with mentors, adjust project plan based on findings, and begin documentation of the CUDA integration process and initial results. Submit midterm evaluation.
   * **Week 7**: Assuming satisfactory progress, begin extending CUDA optimizations to additional DIPY algorithms, such as fiber tracking or segmentation, applying best practices learned from the non-rigid registration work.
   * **Week 8 - 9**: Develop and optimize CUDA implementations for the selected additional algorithms, focusing on achieving significant performance gains.
   * **Week 10**: Conduct comprehensive testing of all CUDA-optimized algorithms, focusing on performance, accuracy, and usability across a variety of use cases.
   * **Week 11**: Finalize all CUDA optimizations, focusing on code quality, documentation, and preparing for a "code freeze" to finalize all changes.
   * **Week 12**: Complete comprehensive documentation of all CUDA-optimized algorithms, including performance benchmarks, best practices, and integration instructions.
   * **Final week**: Submit the project, ensuring all CUDA-optimized algorithms are integrated, well-documented, and ready for deployment within the DIPY ecosystem.

**Reply by pjsjongsung** [reply]
I would mention the other pull request you are working on as well. I also suggest adding more description to the description. Why is non rigid registration crucial? Why does CUDA acceleration help reduce the processing time? etc. Also have a look at cudipy, since that could be your primary package to contribute to.

**Reply by deka27** [reply]
Thank you for the quick reply, I will make the necessary changes and modify.

**Comment by deka27** [comment]
Hi @pjsjongsung,

I will be dropping my application for Project 2 and instead be applying for project 6. Please let me know if its okay to close the discussion or if you have any advice.

**Reply by pjsjongsung** [reply]
Sure. Closing it now.
